packages/multichain/src/middlewares/multichainMethodCallValidator.ts

```diff
@@ -0,0 +1,122 @@
+import { MultiChainOpenRPCDocument } from '@metamask/api-specs';
+import { createAsyncMiddleware } from '@metamask/json-rpc-engine';
+import { rpcErrors } from '@metamask/rpc-errors';
+import { isObject } from '@metamask/utils';
+import type { JsonRpcError, JsonRpcParams } from '@metamask/utils';
+import type {
+  ContentDescriptorObject,
+  MethodObject,
+  OpenrpcDocument,
+  ReferenceObject,
+} from '@open-rpc/meta-schema';
+import dereferenceDocument from '@open-rpc/schema-utils-js/build/dereference-document';
+import { makeCustomResolver } from '@open-rpc/schema-utils-js/build/parse-open-rpc-document';
+import type { Schema, ValidationError } from 'jsonschema';
+import { Validator } from 'jsonschema';
+
+const transformError = (
+  error: ValidationError,
+  param: ContentDescriptorObject,
+  got: unknown,
+) => {
+  // if there is a path, add it to the message
+  const message = `${param.name}${
+    error.path.length > 0 ? `.${error.path.join('.')}` : ''
+  } ${error.message}`;
+
+  return {
+    code: -32602, // TODO: could be a different error code or not wrapped in json-rpc error, since this will also be wrapped in a -32602 invalid params error
+    message,
+    data: {
+      param: param.name,
+      path: error.path,
+      schema: error.schema,
+      got,
+    },
+  };
+};
+
+const v = new Validator();
+
+const dereffedPromise = dereferenceDocument(
+  MultiChainOpenRPCDocument as unknown as OpenrpcDocument,
+  makeCustomResolver({}),
+);
+
+/**
+ * Helper that utilizes the Multichain method specifications from `@metamask/api-specs`
+ * to validate the params of a Multichain request.
+ *
+ * @param method - The request's method.
+ * @param params - The request's optional JsonRpcParams object.
+ * @returns an array of error objects for each validation error or an empty array if no errors.
+ */
+const multichainMethodCallValidator = async (
+  method: string,
+  params: JsonRpcParams | undefined,
+) => {
+  const dereffed = await dereffedPromise;
+
+  const methodToCheck = dereffed.methods.find(
+    (m: MethodObject | ReferenceObject) => (m as MethodObject).name === method,
+  ) as MethodObject | undefined;
+
+  if (
+    !methodToCheck ||
+    !isObject(methodToCheck) ||
+    !('params' in methodToCheck)
+  ) {
+    return [
+      {
+        code: -32601,
```

Comment:
Can we use `rpcErrors` so we use the standard codes, error messages etc and don't have to duplicate those?

Reply:
pushing a PR to address this

Reply:
https://github.com/MetaMask/core/pull/5131